### PR TITLE
refactor: use template literal types for terser code

### PR DESCRIPTION
**Pull request by rabelloo**
## Purpose

With newly introduced [template literal types](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#template-literal-types) we can write terser code.

## Approach

Reuse type variants as much as possible in `color` and `font`.

## Testing

CI checks.

## Risks

Highly subjective to personal preference.

Note: the outcome is still the same, hovering the type name shows the same type union.

### CODE REVIEW

File: packages/core/src/helpers/font/font.ts [hunk 1]
```diff
@@ -41,37 +41,19 @@ export function font(name: FontName): Font {
...
-  | '200-mono';
+  | `${FontSize}-bold`
+  | `${FontSize}-regular`
+  | `${LargeFont}-light`
+  | `${SmallFont}-allcaps`
+  | `${Exclude<SmallFont, '100'>}-mono`;
+
+type FontSize = keyof typeof sizes;
+type LargeFont = Exclude<FontSize, '400' | '300' | '200' | '100'>;
+type SmallFont = Exclude<FontSize, LargeFont>;
```
Comment (josokinas) [inline packages/core/src/helpers/font/font.ts:56]: At first glance this seems a little complex.

What do you think about such, even if it's longer:

```suggestion
  | `${FontSize}-bold`
  | `${FontSize}-regular`
  | `${Exclude<FontSize, '800' | '700' | '600' | '500'>}-light`
  | `${Exclude<FontSize, '400' | '300' | '200' | '100'>}-allcaps`
  | `${Exclude<FontSize, '400' | '300' | '200'>}-mono`;

type FontSize = keyof typeof sizes;
```

File: packages/core/src/helpers/color/color.ts [hunk 2]
```diff
@@ -38,88 +38,38 @@ export type Palette =
...
-  | 'neutral-800'
-  | 'neutral-700'
-  | 'neutral-600'
-  | 'neutral-500'
-  | 'neutral-400'
-  | 'neutral-300'
-  | 'neutral-200'
-  | 'neutral-050'
-  | 'neutral-white'
-  | 'neutral-black';
```
Comment (josokinas) [inline packages/core/src/helpers/color/color.ts:52]: This all is much more difficult to read now...

What do you think if only like such would be a change:

```ts
type NeutralColor =   `neutral-${'900' | '800' | '700' | '600' | '500' | '400' | '300' | '200' | '050' | 'white' | 'black'}`;
```

Or use the `Weight` type, and exclude what's not needed for some cases?

**Review by josokinas — APPROVED** [review]

File: packages/core/src/helpers/color/color.ts (hunk 2 above)
Comment (rabelloo) [reply packages/core/src/helpers/color/color.ts:52]: Fair enough. I reused types as most as I could, but it does make it slightly more annoying to read.

How does it look now? Want me to get rid of `Muted` and `Vivid` too?

Comment (rabelloo) [reply packages/core/src/helpers/color/color.ts:52]: Don't mind it, though I feel it makes it a little less readable, take a look.

Comment (josokinas) [reply packages/core/src/helpers/color/color.ts:52]: So I feel like such we now have only one place where available weight for colors is defined, and others only extract or exclude what they need.

Comment (rabelloo) [reply packages/core/src/helpers/color/color.ts:52]: Cool, unfortunately `Extract` does not suggest keys in the union so it's a bit iffy, but agree.
I'll merge like it is and we can always revisit it if we don't like it.

**Comment by rabelloo** [comment]
Btw if you have ideas to use more of this feature in other places that'd be awesome.

**Pull request merged**